probe/pkg/runtime/runtime.go

```diff
@@ -0,0 +1,249 @@
+}
+
+// New creates a new runtime.
+func New(config *config.Config, fleetManagerClient fleetmanager.Client) (*Runtime, error) {
+	centralName, err := newCentralName()
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to create runtime")
+	}
+
+	request := &public.CentralRequestPayload{
+		Name:          centralName,
+		MultiAz:       true,
+		CloudProvider: config.DataCloudProvider,
+		Region:        config.DataPlaneRegion,
+	}
+
+	return &Runtime{
+		config:             config,
+		fleetManagerClient: fleetManagerClient,
+		request:            request,
+	}, nil
+}
+
+// Run sets up time outs and metrics, and then executes either a single probe run
+// or a continuous loop of probe runs.
+func (r *Runtime) Run(isErr chan bool, sigs chan os.Signal, loop bool) {
+	for {
+		ctxTimeout, cancel := context.WithTimeout(context.Background(), r.config.RuntimeRunTimeout)
+		defer cancel()
+		isDone := make(chan bool, 1)
+
+		go func() {
+			metrics.MetricsInstance().IncStartedRuns()
+			err := r.probeCentral(ctxTimeout)
+			if err != nil {
+				metrics.MetricsInstance().IncFailedRuns()
+				metrics.MetricsInstance().SetLastFailureTimestamp()
+				glog.Error(err)
+			} else {
+				metrics.MetricsInstance().IncSuccessfulRuns()
+				metrics.MetricsInstance().SetLastSuccessTimestamp()
+			}
+			isDone <- true
+		}()
+
+		select {
+		case <-ctxTimeout.Done():
+			glog.Errorf("Probe run timed out: %v", ctxTimeout.Err())
+			metrics.MetricsInstance().IncFailedRuns()
+			if loop {
+				time.Sleep(r.config.RuntimeRunWaitPeriod)
+				continue
+			} else {
+				isErr <- true
+				close(sigs)
+				return
+			}
+		case <-isDone:
+			if loop {
+				time.Sleep(r.config.RuntimeRunWaitPeriod)
+				continue
+			} else {
+				isErr <- false
+				close(sigs)
+				return
+			}
+		}
+	}
+}
+
+// probeCentral represents a single probe run.
+func (r *Runtime) probeCentral(ctx context.Context) error {
+	defer totalDurationTimer()()
+
+	err := r.createCentral(ctx)
+	if err != nil {
+		return err
+	}
+
+	err = r.verifyCentral(ctx)
+	if err != nil {
+		return err
+	}
+
+	err = r.deleteCentral(ctx)
+	return err
+}
+
+// Stop the probe run.
+// TODO: Add write ahead log to clean up after ungraceful restarts.
+func (r *Runtime) Stop() error {
+	if r.centralState == nil {
+		return nil
+	}
+	ctx := context.Background()
+	_, err := r.fleetManagerClient.DeleteCentralById(ctx, r.centralState.Id, true)
+	if err != nil {
+		return errors.Wrapf(err, "Clean up: Deletion of Central instance %s failed", r.centralState.Id)
+	}
+	glog.Infof("Clean up: Deletion of Central instance %s requested.", r.centralState.Id)
+	return nil
+}
+
+// totalDurationTimer returns a function that measures the elapsed time between the
+// call to the timer and the call to the returned function. The result is tracked
+// as a Prometheus metric.
+func totalDurationTimer() func() {
+	start := time.Now()
+	return func() {
+		metrics.MetricsInstance().ObserveTotalDuration(time.Since(start))
+	}
+}
```

Comment:
Reading the callsite `defer totalDurationTimer()()` makes it hard to see what's going on, both because the name isn't very descriptive, and also because of the double function call. I would propose to change it to
```suggestion
func recordElapsedTime(start time.Time) {
	metrics.MetricsInstance().ObserveTotalDuration(time.Since(start))
}
```
an then call it via `defer recordElapsedTime(time.Now())` (note that arguments to `defer`ed functions are evaluated eagerly)